src/coff_import_file.rs

```diff
@@ -3,7 +3,898 @@
+    // string content itself, which is an array of null-terminated C-style
+    // strings.  The termination is important as they are referenced to by offset
+    // by the symbol entity in the file format.
+
+    let offset = b.len();
+
+    // Skip over the length field, we will fill it in later as we will have
+    // computed the length while emitting the string content itself.
+    b.resize(offset + size_of::<u32>(), 0);
+
+    for s in strings {
+        b.write_all(s)?;
+        b.write_all(&[0])?;
+    }
+
+    // Backfill the length of the table now that it has been computed.
+    let length: u32 = (b.len() - offset).try_into().unwrap();
+    b[offset..offset + size_of::<u32>()].copy_from_slice(&length.to_le_bytes());
+
+    Ok(())
+}
+
+fn get_name_type(sym: &str, ext_name: &str, machine: MachineTypes, mingw: bool) -> ImportNameType {
+    // A decorated stdcall function in MSVC is exported with the
+    // type IMPORT_NAME, and the exported function name includes the
+    // the leading underscore. In MinGW on the other hand, a decorated
+    // stdcall function still omits the underscore (IMPORT_NAME_NOPREFIX).
+    // See the comment in isDecorated in COFFModuleDefinition.cpp for more
+    // details.
+    if ext_name.starts_with('_') && ext_name.contains('@') && !mingw {
+        ImportNameType::Name
+    } else if sym != ext_name {
+        ImportNameType::NameUndecorate
+    } else if machine == MachineTypes::I386 && sym.starts_with('_') {
+        ImportNameType::NameNoprefix
+    } else {
+        ImportNameType::Name
+    }
+}
+
+fn replace(s: &str, mut from: &str, mut to: &str) -> Result<String> {
+    if let Some((before, after)) = s.split_once(from) {
+        return Ok(format!("{before}{to}{after}"));
+    }
+
+    // From and To may be mangled, but substrings in S may not.
+    if from.starts_with('_') && to.starts_with('_') {
+        from = &from[1..];
+        to = &to[1..];
+        if let Some((before, after)) = s.split_once(from) {
+            return Ok(format!("{before}{to}{after}"));
+        }
+    }
+
+    Err(Error::other(format!(
+        "{s}: replacing '{from}' with '{to}' failed"
+    )))
+}
+
+/// This class constructs various small object files necessary to support linking
+/// symbols imported from a DLL.  The contents are pretty strictly defined and
+/// nearly entirely static.  The details of the structures files are defined in
+/// WINNT.h and the PE/COFF specification.
+struct ObjectFactory<'a> {
+    native_machine: MachineTypes,
+    import_name: &'a str,
+    import_descriptor_symbol_name: Vec<u8>,
+    null_thunk_symbol_name: Vec<u8>,
+}
+
+unsafe fn any_as_u8_slice<T: Sized>(p: &T) -> &[u8] {
+    slice::from_raw_parts((p as *const T) as *const u8, size_of::<T>())
+}
+
+impl<'a> ObjectFactory<'a> {
+    fn new(s: &'a str, m: MachineTypes) -> Self {
+        let import_as_path = PathBuf::from(s);
+        let library = import_as_path.file_stem().unwrap().to_string_lossy();
```

Comment:
This would silently produce a wrong import library if the file stem is not UTF-8. Can this be made to return an error instead?